### PR TITLE
Introduce Puppeteer for handling browsers during tests

### LINKED ISSUE #11628: Fix or switch test harness
On the linux bot I see lots of test-runs that have never finished with a lot of chrome processes open still. Chrome also seems to stay open on windows sometimes as well.

I'm starting to think we should consider switching our test runner to puppeteer or something similar that handles starting up and shutting down the browser.

### PR DESCRIPTION
**Pull request by timvandermeij**
The commit messages contain more information about the individual changes.

Fixes #11628.

### CODE REVIEW

**Comment by timvandermeij** [comment]
Well, at least one of the bots is happy. The Windows bot was already broken, so perhaps that fixes itself after a gentle kick.

**Comment by timvandermeij** [comment]
The browser versions are most likely different from those installed on the bots, so I'm expecting some changes in the reference images here.

**Comment by timvandermeij** [comment]
I'm pleasantly surprised that this is working immediately on the Linux bot. It looks like I broke the Windows bot even more, but I'll contact Brendan to get that fixed so we can also trigger the tests there to see how that goes.

**Comment by Snuffleupagus** [comment]
This looks like a very good improvement, and the amount of code (or lack thereof) that was required to get this to work is a pleasant surprise!

A couple of questions and things I noticed (based on a quick look):
 - > Moreover, this ensures that contributors always run the tests in both Firefox and Chrome, regardless of which browsers they have installed locally.

   This obviously makes a lot of sense as a default, but could we perhaps add a command-line flag (e.g. `--skip-chromium`) for the test tasks such that there's an easy way to skip Chrome; please?
   To save time/resources, I'm always testing only in Firefox locally (via a suitable browser_manifest.json file) and it would be absolutely great to still be able to do that without having to edit the test files. 

 - Looking at the bot logs, the unit- and font-tests still print their output as before, e.g.
    ```
    Run 1296 tests
    All unit tests passed.
    unit tests runtime was 134.1 seconds
    ```
   However, the `browsertest` no longer appears to log it's success/failure output as before.[1]

---
[1] I'd expect something *similar* to e.g. 
```
OHNOES!  Some tests failed!
  different ref/snapshot: 8
  different first/second rendering: 2
Runtime was 989.1 seconds
[11:41:37] Finished '<anonymous>' after 19 min
[11:41:37] Finished 'bottest' after 19 min
```

**Comment by timvandermeij** [comment]
Thank you for the feedback!

> could we perhaps add a command-line flag (e.g. --skip-chromium) for the test tasks such that there's an easy way to skip Chrome

Yes, that sounds like a good idea. I have added it to the to do list.

> However, the browsertest no longer appears to log it's success/failure output as before

That's a good point, I totally missed that. I have no idea where it went, but I'll look into it; it's also on the list now.

**Comment by Snuffleupagus** [comment]
One more thing, does benchmarking[1] still work as well? I'd expect the answer to be yes, since everything else seem to work nicely, but please test and/or confirm that that's the case :-)

---
[1] Following the instructions at https://github.com/mozilla/pdf.js/wiki/Benchmarking-your-changes

**Comment by timvandermeij** [comment]
This is really strange. The reason why the summary logs for the reference tests aren't there is because the test process apparently stops during the reference test run. However, this does not happen locally at all: all tests are run the summary is logged. I have added extensive logging, but it just stops: no closing logic is called whatsoever...

**Comment by Snuffleupagus** [comment]
Wait, is this perhaps simply running into the *limited* log size available on the bots!?
Between the "Extracting..." and "TEST-UNEXPECTED-FAIL | ..." lines, the logs are a lot larger than usual.

**Comment by timvandermeij** [comment]
Do you know of a log size limit that we configured somewhere? I'll try to get rid of the "Extracting..." lines to see if that's indeed the case.

**Comment by Snuffleupagus** [comment]
> Do you know of a log size limit that we configured somewhere?

No idea unfortunately, but in any case that probably requires access to the bots...

> I'll try to get rid of the "Extracting..." lines to see if that's indeed the case.

If I were to guess, I'd say that the sheer volume of reference test failures might be the biggest problem w.r.t. to the log size. That hypothesis should be easy enough to test though, try running `makeref` and then re-run the `test` afterwards :-)
(Obviously you'd then have to re-run `makeref` against e.g. master afterwards, to restore things to a working state.)

**Comment by timvandermeij** [comment]
I have implemented the "Extracting..." log suppression, so let's see if we get further now.

/botio-linux test

**Comment by timvandermeij** [comment]
@Snuffleupagus I think you're absolutely right. Having removed the "Extracting..." logs made it go much further. I looked into how such a limit would be configured, but I can't find anything about it. Maybe Node.js decides to kill the process once it reaches a certain limit or something. Anyway, thanks for thinking along since I was a bit stuck here!

Next thing I'm going to try is to use exactly the same versions as currently installed on the bots in the hope to get rid of the failures.

File: test/test.js [hunk 1]
```diff
@@ -784,52 +760,102 @@ function unitTestPostHandler(req, res) {
...
-      browser: browser,
-      closed: false,
+
+  function filterLogs() {
+    // Hook into the standard output to remove extraction logs from Puppeteer
+    // since they spam the logs and are not useful for us.
+    const oldWrite = process.stdout.write;
+    process.stdout.write = (function(write) {
+      return function(string, encoding, fd) {
+        if (string.includes("Extracting...")) {
```
[the remark this answers is not in the document]
Comment (timvandermeij) [reply test/test.js]: Done in the new commit.

File: test/test.js [hunk 2]
```diff
@@ -784,52 +760,102 @@ function unitTestPostHandler(req, res) {
...
+          return;
+        }
+        write.apply(process.stdout, [string, encoding, fd]);
```
[the remark this answers is not in the document]
Comment (timvandermeij) [reply test/test.js]: Done in the new commit.

**Comment by brendandahl** [comment]
Hmm...looks like firefox stays open on the windows bot and I see:
```
Mon Apr 13 2020 17:56:00 GMT+0000 (Greenwich Mean Time): Command queued at #0 ("cmd_test" from @timvandermeij, id: 7620f52af18fa63)
shell.js: internal error
Error: EPERM: operation not permitted, unlink 'C:\pdfjs\botio-files-pdfjs\private\7620f52af18fa63/node_modules/puppeteer/.local-firefox/win64-76/firefox/browser/features/webcompat@mozilla.org.xpi'
    at Object.unlinkSync (fs.js:956:3)
    at _unlinkSync (C:\pdfjs\botio\node_modules\shelljs\shell.js:1398:10)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1145:9)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
```

**Comment by timvandermeij** [comment]
I have amended the commit to add a flag that Windows seems to require to be able to start the browser properly, as well as exception handling for when the browser cannot be launched so the process continues nicely instead of failing due to an uncaught promise rejection (causing the bots to wait forever).

**Comment by timvandermeij** [comment]
Cool, it's working on Windows now as well. I only have to fix the unhandled promise rejection at the end now, but fortunately it's not fatal and the processes stopped nicely. I have updated the list above.

File: gulpfile.js [hunk 3]
```diff
@@ -436,6 +436,9 @@ function createTestSource(testsName, bot) {
     if (bot) {
       args.push("--strictVerify");
     }
+    if (process.argv.includes("--noChrome")) {
+      args.push("--noChrome");
+    }
```
Comment (Snuffleupagus) [inline gulpfile.js:441]: As I hoped, when asking for this feature, it didn't require a lot of code; thanks for doing this!

Just one small thing: Can you please add this same block to the `makeRef` function below as well, such that *all* the various test commands are covered?

Comment (timvandermeij) [reply gulpfile.js:441]: Done in the new commit series.

**Comment by timvandermeij** [comment]
This should be ready for prime time now. I have updated the PR description with which upstream issues exist, but they are no show stoppers for us because we have workarounds in place. They are simply filed to improve the upstream code and allow us to remove workarounds once they get fixed upstream.

@Snuffleupagus Would you be willing to review this?

**Review by Snuffleupagus — APPROVED** [review]
The various test-commands still seem to work just fine locally on Windows, which I believe was the most important part for me to check here :-)
Thanks for putting in the, clearly not small, amount of work required to make this happen!

The only slight annoyance I found is that the elements containing logs and ref-test images are now a lot smaller than previously, and doesn't seem to auto-expand with the window size, to the point of being *very* difficult to use.
 - The following DOME element is smaller than previously (during test runs): https://github.com/mozilla/pdf.js/blob/master/test/test_slave.html#L31
 - The DOM element containing the ref-test image comparison is a lot smaller (requiring too much scrolling when comparing failures): https://github.com/mozilla/pdf.js/blob/master/test/resources/reftest-analyzer.html#L53

I suppose that this could be Windows specific, but if there's a simple fix for this that'd be great!
(This is the only thing that I'd *really* appreciate being fixed before this PR lands.)

---

I've not reviewed this in too much detail, since I see the value of actually getting this landed without further delays. (If there's further improvements necessary, they can simply be done in follow-ups.)

**Comment by timvandermeij** [comment]
I have fixed the too small elements. It turns out by default Puppeteer uses an 800 x 600 viewport. This is configurable and I have now disabled that to make all elements render normally.

/botio makeref

**Pull request merged**